[Bug]: AMI Distribution Configuration block in Image Builder Distribution Configuration uses incorrect regular expression to validate output AMI name

### Terraform Core Version

1.5.5

### AWS Provider Version

5.42.0,5.41.0,5.40.0

### Affected Resource(s)

* aws_imagebuilder_distribution_configuration

### Expected Behavior

The output AMI name (`aws_imagebuilder_distribution_configuration.*.ami_distribution_configuration.*.name`) allows literal periods (`.`) per the [AmiDistributionConfiguration specification](https://docs.aws.amazon.com/imagebuilder/latest/APIReference/API_AmiDistributionConfiguration.html).

Example: `AmazonLinux2-EKS-1.27` is a valid value to supply for the above named attribute, and executing `terraform apply` creates the resource with that as the output AMI name. 

### Actual Behavior

When setting `name = "AmazonLinux2-EKS-1.27"`, we get the error message below, and the resource does not get created.



### Relevant Error/Panic Output Snippet

```shell
╷
│ Error: invalid value for distribution.0.ami_distribution_configuration.0.name (must contain only alphanumeric characters, underscores, and hyphens)
│ 
│   with aws_imagebuilder_distribution_configuration.distribution_configuration,
│   on example.tf line 12, in resource "aws_imagebuilder_distribution_configuration" "distribution_configuration":
│   12: resource "aws_imagebuilder_distribution_configuration" "distribution_configuration" {
│
```


### Terraform Configuration Files

```hcl
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "= 5.40.0"
    }
  }

  required_version = "1.5.5"
}

resource "aws_imagebuilder_distribution_configuration" "distribution_configuration" {
  name = "AmazonLinux2-EKS-1-27"
  distribution {
    region = "us-east-1"
    ami_distribution_configuration {
      name = "AmazonLinux2-EKS-1.27"
    }
  }
}

```

### Steps to Reproduce

Prerequisite: Terraform 1.5.5

1. Create a new directory
2. Copy the above text into a file called `example.tf`
3. Execute `terraform init`
4. Execute `terraform plan`

The plan will fail with the above error.

### Debug Output

```
2024-03-27T15:06:06.925Z [INFO]  Terraform version: 1.5.5
2024-03-27T15:06:06.925Z [DEBUG] using github.com/hashicorp/go-tfe v1.26.0
2024-03-27T15:06:06.925Z [DEBUG] using github.com/hashicorp/hcl/v2 v2.16.2
2024-03-27T15:06:06.926Z [DEBUG] using github.com/hashicorp/terraform-svchost v0.1.0
2024-03-27T15:06:06.926Z [DEBUG] using github.com/zclconf/go-cty v1.12.2
2024-03-27T15:06:06.926Z [INFO]  Go runtime version: go1.20.7
2024-03-27T15:06:06.926Z [INFO]  CLI args: []string{"/opt/tfenv/versions/1.5.5/terraform", "plan"}
2024-03-27T15:06:06.926Z [DEBUG] Attempting to open CLI config file: /home/redacted/.terraformrc
2024-03-27T15:06:06.926Z [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2024-03-27T15:06:06.927Z [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2024-03-27T15:06:06.927Z [DEBUG] ignoring non-existing provider search directory /home/redacted/.terraform.d/plugins
2024-03-27T15:06:06.927Z [DEBUG] ignoring non-existing provider search directory /home/redacted/.local/share/terraform/plugins
2024-03-27T15:06:06.927Z [DEBUG] ignoring non-existing provider search directory /usr/local/share/terraform/plugins
2024-03-27T15:06:06.927Z [DEBUG] ignoring non-existing provider search directory /usr/share/terraform/plugins
2024-03-27T15:06:06.929Z [INFO]  CLI command args: []string{"plan"}
2024-03-27T15:06:06.936Z [DEBUG] New state was assigned lineage "redacted"
2024-03-27T15:06:08.574Z [DEBUG] checking for provisioner in "."
2024-03-27T15:06:08.574Z [DEBUG] checking for provisioner in "/opt/tfenv/versions/1.5.5"
2024-03-27T15:06:08.578Z [INFO]  backend/local: starting Plan operation
2024-03-27T15:06:08.592Z [DEBUG] created provider logger: level=debug
2024-03-27T15:06:08.592Z [INFO]  provider: configuring client automatic mTLS
2024-03-27T15:06:08.622Z [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/hashicorp/aws/5.40.0/linux_amd64/terraform-provider-aws_v5.40.0_x5 args=[.terraform/providers/registry.terraform.io/hashicorp/aws/5.40.0/linux_amd64/terraform-provider-aws_v5.40.0_x5]
2024-03-27T15:06:08.686Z [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/hashicorp/aws/5.40.0/linux_amd64/terraform-provider-aws_v5.40.0_x5 pid=147124
2024-03-27T15:06:08.686Z [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/hashicorp/aws/5.40.0/linux_amd64/terraform-provider-aws_v5.40.0_x5
2024-03-27T15:06:09.399Z [INFO]  provider.terraform-provider-aws_v5.40.0_x5: configuring server automatic mTLS: timestamp=2024-03-27T15:06:09.398Z
2024-03-27T15:06:09.449Z [DEBUG] provider: using plugin: version=5
2024-03-27T15:06:09.449Z [DEBUG] provider.terraform-provider-aws_v5.40.0_x5: plugin address: address=/tmp/plugin2403709041 network=unix timestamp=2024-03-27T15:06:09.448Z
2024-03-27T15:06:11.033Z [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2024-03-27T15:06:11.046Z [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/hashicorp/aws/5.40.0/linux_amd64/terraform-provider-aws_v5.40.0_x5 pid=147124
2024-03-27T15:06:11.046Z [DEBUG] provider: plugin exited
2024-03-27T15:06:11.047Z [DEBUG] Building and walking validate graph
2024-03-27T15:06:11.048Z [DEBUG] ProviderTransformer: "aws_imagebuilder_distribution_configuration.distribution_configuration" (*terraform.NodeValidatableResource) needs provider["registry.terraform.io/hashicorp/aws"]
2024-03-27T15:06:11.049Z [DEBUG] ReferenceTransformer: "provider[\"registry.terraform.io/hashicorp/aws\"]" references: []
2024-03-27T15:06:11.049Z [DEBUG] ReferenceTransformer: "aws_imagebuilder_distribution_configuration.distribution_configuration" references: []
2024-03-27T15:06:11.050Z [DEBUG] Starting graph walk: walkValidate
2024-03-27T15:06:11.056Z [DEBUG] created provider logger: level=debug
2024-03-27T15:06:11.057Z [INFO]  provider: configuring client automatic mTLS
2024-03-27T15:06:11.072Z [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/hashicorp/aws/5.40.0/linux_amd64/terraform-provider-aws_v5.40.0_x5 args=[.terraform/providers/registry.terraform.io/hashicorp/aws/5.40.0/linux_amd64/terraform-provider-aws_v5.40.0_x5]
2024-03-27T15:06:11.097Z [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/hashicorp/aws/5.40.0/linux_amd64/terraform-provider-aws_v5.40.0_x5 pid=147148
2024-03-27T15:06:11.097Z [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/hashicorp/aws/5.40.0/linux_amd64/terraform-provider-aws_v5.40.0_x5
2024-03-27T15:06:11.766Z [INFO]  provider.terraform-provider-aws_v5.40.0_x5: configuring server automatic mTLS: timestamp=2024-03-27T15:06:11.766Z
2024-03-27T15:06:11.788Z [DEBUG] provider.terraform-provider-aws_v5.40.0_x5: plugin address: address=/tmp/plugin3594176402 network=unix timestamp=2024-03-27T15:06:11.788Z
2024-03-27T15:06:11.788Z [DEBUG] provider: using plugin: version=5
2024-03-27T15:06:12.886Z [WARN]  provider.terraform-provider-aws_v5.40.0_x5: [WARN] Truncating attribute path of 1 diagnostics for TypeSet
2024-03-27T15:06:12.887Z [ERROR] provider.terraform-provider-aws_v5.40.0_x5: Response contains error diagnostic: diagnostic_severity=ERROR tf_proto_version=5.4 tf_provider_addr=registry.terraform.io/hashicorp/aws tf_rpc=ValidateResourceTypeConfig @module=sdk.proto diagnostic_attribute=AttributeName("distribution") diagnostic_detail= tf_req_id=86ac0b73-59dd-35fd-067b-c9ea3eeb231f @caller=github.com/hashicorp/terraform-plugin-go@v0.22.0/tfprotov5/internal/diag/diagnostics.go:58 diagnostic_summary="invalid value for distribution.0.ami_distribution_configuration.0.name (must contain only alphanumeric characters, underscores, and hyphens)" tf_resource_type=aws_imagebuilder_distribution_configuration timestamp=2024-03-27T15:06:12.886Z
2024-03-27T15:06:12.888Z [ERROR] vertex "aws_imagebuilder_distribution_configuration.distribution_configuration" error: invalid value for distribution.0.ami_distribution_configuration.0.name (must contain only alphanumeric characters, underscores, and hyphens)
╷
│ Error: invalid value for distribution.0.ami_distribution_configuration.0.name (must contain only alphanumeric characters, underscores, and hyphens)
│ 
│   with aws_imagebuilder_distribution_configuration.distribution_configuration,
│   on example.tf line 12, in resource "aws_imagebuilder_distribution_configuration" "distribution_configuration":
│   12: resource "aws_imagebuilder_distribution_configuration" "distribution_configuration" {
│ 
╵
2024-03-27T15:06:12.893Z [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2024-03-27T15:06:12.910Z [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/hashicorp/aws/5.40.0/linux_amd64/terraform-provider-aws_v5.40.0_x5 pid=147148
2024-03-27T15:06:12.910Z [DEBUG] provider: plugin exited
```

### Panic Output

No panic occurred.

### Important Factoids

_No response_

### References

[AmiDistributionConfiguration specification](https://docs.aws.amazon.com/imagebuilder/latest/APIReference/API_AmiDistributionConfiguration.html)

### Would you like to implement a fix?

No

--- a/.changelog/36659.txt
+++ b/.changelog/36659.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_imagebuilder_distribution_configuration: Fix validation regex for `ami_distribution_configuration.name`
+```

--- a/internal/service/imagebuilder/distribution_configuration.go
+++ b/internal/service/imagebuilder/distribution_configuration.go
@@ -121,7 +121,7 @@ func ResourceDistributionConfiguration() *schema.Resource {
 										Optional: true,
 										ValidateFunc: validation.All(
 											validation.StringLenBetween(0, 127),
-											validation.StringMatch(regexache.MustCompile(`^[0-9A-Za-z_{-][0-9A-Za-z_\s:{}-]+[0-9A-Za-z_}-]$`), "must contain only alphanumeric characters, underscores, and hyphens"),
+											validation.StringMatch(regexache.MustCompile(`^[0-9A-Za-z_{-][0-9A-Za-z_\.\s:{}-]+[0-9A-Za-z_}-]$`), "must be a valid output AMI name"),
 										),
 									},
 									"target_account_ids": {

--- a/internal/service/imagebuilder/distribution_configuration_test.go
+++ b/internal/service/imagebuilder/distribution_configuration_test.go
@@ -447,6 +447,18 @@ func TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_na
 					}),
 				),
 			},
+			{
+				Config: testAccDistributionConfigurationConfig_amiName(rName, "AmazonLinux2-EKS-1.27-{{ imagebuilder:buildDate }}"),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckDistributionConfigurationExists(ctx, resourceName),
+					acctest.CheckResourceAttrRFC3339(resourceName, "date_updated"),
+					resource.TestCheckResourceAttr(resourceName, "distribution.#", "1"),
+					resource.TestCheckTypeSetElemNestedAttrs(resourceName, "distribution.*", map[string]string{
+						"ami_distribution_configuration.#":      "1",
+						"ami_distribution_configuration.0.name": "AmazonLinux2-EKS-1.27-{{ imagebuilder:buildDate }}",
+					}),
+				),
+			},
 		},
 	})
 }

